### ISSUE TITLE
Unexpected High PCIe traffic in Distributed Training since PT 2

### ISSUE THREAD

**Issue by lchu6** [pytorch/pytorch#103254]
### 🐛 Describe the bug

Since PT 2, we have noticed significant amount of PCIe traffic between host and device, which is something we didn't expect to happen and not observed in PT 1.x version. This applies to both DDP and FSDP, and we observed as much as 2G/s to 4G/s traffic throughout of the whole training stage (for our jobs with 3B+ models) which eventually harm the training due to continuous high consumption of the PCIe and the host.

To reproduce, I am using the same code from the official DDP tutorial: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#initialize-ddp-with-torch-distributed-run-torchrun
except:
1. modify model size to test different model sizes
2. add a training loop to observe the continuous traffic
```python
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim

from torch.nn.parallel import DistributedDataParallel as DDP

from tqdm import tqdm

class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 100000000)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(100000000, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


def demo_basic():
    dist.init_process_group("nccl")
    rank = dist.get_rank()
    print(f"Start running basic DDP example on rank {rank}.")

    # create model and move it to GPU with id rank
    device_id = rank % torch.cuda.device_count()
    model = ToyModel().to(device_id)
    print(f"\n--> model has {sum(p.numel() for p in model.parameters() if p.requires_grad)/1e6} Million params\n")
    ddp_model = DDP(model, device_ids=[device_id])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)

    for _ in tqdm(range(1000000000)):
        optimizer.zero_grad()
        outputs = ddp_model(torch.randn(20, 10))
        labels = torch.randn(20, 5).to(device_id)
        loss_fn(outputs, labels).backward()
        optimizer.step()
```

PCIe traffic:

  | Prior 2.0 (1.13.1+cu117) | Prior 2.0 (1.13.1+cu117) | 2.x (2.1.0.dev20230603+cu118) | 2.x (2.1.0.dev20230603+cu118)
-- | -- | -- | -- | --
model size | device to host (DtoH) | host to device (HtoD) | device to host (DtoH) | host to device (HtoD)
1.6B | 8.8M/s | 21M/s | 122M/s | 950M/s
160M | 10.8M/s | 24.5M/s | 122M/s | 950M/s
16M | 21M/s | 48M/s | 73M/s | 480M/s
1.6M | 9M/s | 45M/s | 25M/s | 170M/s
0.16M | 8.5M/s | 41.5M/s | 7M/s | 36M/s

Some screenshots (for 1.6B model):
PT 1.x
<img width="1692" alt="image" src="https://github.com/pytorch/pytorch/assets/20955448/681a33fe-eb13-4619-9656-27a93ce08f16">
PT 2.x
<img width="1697" alt="image" src="https://github.com/pytorch/pytorch/assets/20955448/bb73ab7a-4e79-4668-909b-b58124ceb7c2">

Some other notes from our experiences:
1. we can confirm the traffic came only from back path (i.e. `loss.backward()`)
2. for DDP, we observed very little to zero PCIe traffic in PT 1.x (as expected), but very high PCIe traffic in PT 2.x
3. for FSDP, we observed high PCIe traffic in PT 1.x as well (which isn't as expected as we didn't turn on any offloading in FSDP), and in PT 2.x, the traffic is even higher (roughly 2X for the same training)
4. We were able to bring PCIe traffic down to almost zero in all these experiments by turning on CUDA_LAUNCH_BLOCKING. However, that is only for debugging purpose as turning on blocking will slow down the training.

### Versions

```
PyTorch version: 2.1.0.dev20230603+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-372.19.1.el8_6.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB
GPU 4: NVIDIA A100-SXM4-80GB
GPU 5: NVIDIA A100-SXM4-80GB
GPU 6: NVIDIA A100-SXM4-80GB
GPU 7: NVIDIA A100-SXM4-80GB

Nvidia driver version: 515.48.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          80
On-line CPU(s) list:             0-79
Thread(s) per core:              2
Core(s) per socket:              20
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel Xeon Processor (Cascadelake)
Stepping:                        5
CPU MHz:                         2400.000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.3 MiB
L1i cache:                       1.3 MiB
L2 cache:                        160 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-39
NUMA node1 CPU(s):               40-79
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat pku ospke avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.25.0rc1
[pip3] pytorch-triton==2.1.0+9820899b38
[pip3] torch==2.1.0.dev20230603+cu118
[pip3] torchvision==0.16.0.dev20230603+cu118
[conda] numpy                     1.25.0rc1                pypi_0    pypi
[conda] pytorch-triton            2.1.0+9820899b38          pypi_0    pypi
[conda] torch                     2.1.0.dev20230603+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230603+cu118          pypi_0    pypi
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @lessw2020 @HamidShojanazeri

**Comment by kumpera** [comment]
Could you share a profiler run?
If the extra BW is due to explicit D2H copies, they will show up in the profiler output and we should be able to figure out why.

**Comment by lchu-ibm** [comment]
@kumpera 
[toy-example-traces.zip](https://github.com/pytorch/pytorch/files/11691903/toy-example-traces.zip)

**Comment by awgu** [comment]
Could you get the trace with `with_stack=True`?

I think we can just see where these `aten::to` are coming from in DDP forward that are causing the host-to-device copy.

**Comment by awgu** [comment]
Nevermind, I did not realize these traces were from the toy example above. The H2D copies should just be from:
```
outputs = ddp_model(torch.randn(20, 10))
labels = torch.randn(20, 5).to(device_id)
```
DDP will move the input to CUDA, and there is an explicit `.to(device_id)` call as well.

**Comment by lchu-ibm** [comment]
@awgu yes, that was our understanding as well.

Also, as I wrote in the note, we could confirm that for all experiments we did (ddp toy, ddp 1B - 10B, fsdp 3B - ~20B), all extra traffic came from backward.  If we disable backward, it gives the expected traffic, which is putting data to CUDA.

**Comment by awgu** [comment]
By the way, how do you measure the PCIe traffic?

**Comment by walkup** [comment]
Some additional data is listed here for a BERT-large training job using PyTorch DDP() and examples from the PyTorch FSDP Workshop (https://github.com/pytorch/workshops/tree/master/FSDP_Workshop) using T5-3B  / T5-11B and FSDP() with the FULL_SHARD strategy.  Most of the measurements were with two nodes, with 8 NVIDIA A100 GPUs per node, so 16 GPUs total, and one network interface per node with NCCL_ALGO=Ring.  With this configuration, processes on GPUs 0 and 1 use the external network, and processes on the remaining six GPUs use NVLink to communicate with their neighbors on the rings.  We monitored PCIE transmit (D2H = device to host) and receive (H2D = host to device) traffic for all GPUs.  For simplicity let’s focus on GPU 3, which participates in the rings via NVLink only. For these experiments, PCIE traffic was monitored with NVIDIA’s “dcgmi” tool. 

For BERT-large training, the time-averaged PCIE traffic for GPU 3 was  D2H = 0.34 GB/sec, H2D = 2.72 GB/sec when using PyTorch 1.13.1 (torch.__version__ = 1.13.1+cu117).  When testing with a recent Pytorch 2 nightly build (2.1.0.dev20230606+cu121) the training rate remained constant while PCIE traffic increased : D2H = 0.36 GB/sec, H2D = 2.94 GB/sec.  We noticed that the ratio H2D / D2H is about 8x in both cases.  

Differences in PCIE traffic were more pronounced for the FSDP test cases.  For our T5-3B setup, the time-averaged PCIE traffic was  D2H = 0.29 GB/sec, H2D = 2.31 GB/sec with PyTorch 1.13.1, while the Pytorch 2 nightly build resulted in D2H = 0.53 GB/sec, H2D = 4.34 GB/sec.  Again the ratio H2D / D2H is about 8x.  For the FSDP test cases, the training rate improved with the latest Pytorch by about 20% due to better overlap of computation and communication on the GPUs, but the PCIE receive traffic increased by a much larger factor, ~1.9x.  The behavior was about the same for T5-11B.  

Additional measurements show that the PCIE traffic depends on the number of network interfaces that participate in NCCL communication.  Roughly speaking, PCIE traffic decreases when more network interfaces are enabled.  If we set CUDA_LAUNCH_BLOCKING=1, there is no overlap of computation and communication on the GPUs, the training rate decreases by less than 2x, while PCIE traffic becomes negligible.  Of course launch blocking is not viable for training jobs, but it might be useful for debugging purposes.

**Comment by walkup** [comment]
![t5-3b-pt2dev-fsdp](https://github.com/pytorch/pytorch/assets/2922068/78a008d0-7430-45fe-bcb4-f60c7de14f53)
![t5-3b-pt1131-fsdp](https://github.com/pytorch/pytorch/assets/2922068/e7e5846a-4419-4b2e-8707-c9c96f63a79c)
Attached are screen shots from pytorch profiles for T5-3B FSDP (FULL_SHARD) using PyTorch 1.13.1 and our Pytorch 2 development version.  The profile for 1.13.1 has compute-stream 7, NCCL stream 84, and you can see little overlap of compute and NCCL.  In contrast, the profile for PyTorch 2 has compute stream 7, NCCL stream 32, and there is an additional stream 28 which calls direct_copy_kernel_cuda every 10 milliseconds.  No changes were made to the job parameters other than to load a different conda environment.  The NCCL versions are different, and the NCCL kernel times are faster with the older NCCL in PyTorch 1.13.1.  The increase in time-averaged PCIE traffic (1.9x) is much more than would result from the increased training rate (1.2x).  The PyTorch profiles when gzipped are about 12 MB each … I don’t know if it is OK to attach them.  I should mention that the PCIE traffic is not present in benchmarks that simply run NCCL in a loop, using either CUDA/NCCL tests or using python and calls such as torch.distributed.all_reduce().

**Comment by awgu** [comment]
Oh ok. I think the improved overlap with PT2 is due to doing the full-to-low precision cast in a separate stream. This makes sense.

**Comment by walkup** [comment]
[t5-3b-pt2dev.pt.trace.json.gz](https://github.com/pytorch/pytorch/files/11708635/t5-3b-pt2dev.pt.trace.json.gz)
[t5-3b-pt1131.pt.trace.json.gz](https://github.com/pytorch/pytorch/files/11708639/t5-3b-pt1131.pt.trace.json.gz)
PyTorch profiles are attached.

**Comment by raghukiran1224** [comment]
> Oh ok. I think the improved overlap with PT2 is due to doing the full-to-low precision cast in a separate stream. This makes sense.

@awgu During the overlap, are there paths that lead to copying of the data from GPU to CPU?

**Comment by lchu-ibm** [comment]
@awgu I am wondering if we have any guess/thought on the potential cause?   I know the discussion above with FSDP perf is also something interesting but the pcie issue appears to happen for all distributed training (ddp, fsdp with perf diff, fsdp with no perf diff, etc.).  So while you guys are still investigating, I am wondering if there is some thing we can also help test.  we have well set up some toy and real models around this issue with clear dashboard on all the traffic, so please let us know if there is any thought/guess that we can help test.

**Comment by awgu** [comment]
I wonder if we can narrow down if the PCIe traffic is coming from collective calls or not. Let me know if there is already evidence against this.

IIUC, the `ToyModel` has 4 `nn.Parameter`s (one `weight` and `bias` for each of two `nn.Linear`s), which should lead to 3-4 all-reduce buckets if using DDP. (I think the last `bias: (5,)` will be pushed to the next bucket, in which case there should be 4 buckets, but I am not 100% sure off the top of my head.)

Could we try to specify a massive `bucket_cap_mb` such that there is only one all-reduce, and could we try to correlate if the PCIe traffic spike corresponds exactly to when that all-reduce is called?

> 1. we can confirm the traffic came only from back path (i.e. loss.backward())

This suggests though that the traffic is not only for the collectives since FSDP runs all-gathers in forward as well. I just wanted to check that the observation that the traffic comes from backward also holds for FSDP?

**Comment by walkup** [comment]
[css-host-180.t5-3b.06_20_13_36_56.nsys-rep.zip](https://github.com/pytorch/pytorch/files/11801493/css-host-180.t5-3b.06_20_13_36_56.nsys-rep.zip)
[dmon.txt](https://github.com/pytorch/pytorch/files/11801494/dmon.txt)
![t5-3b 06_20_13_36_56](https://github.com/pytorch/pytorch/assets/2922068/2d6ef7e6-3fe2-495a-9a6b-0047c7fdb08e)
NVIDIA’s Nsight Systems can sample PCIE traffic at ~0.1 msec intervals along with capturing other metrics and GPU kernels including NCCL calls.  A nsys profile report is attached for the process assigned to GPU 3 on the master node of a two-node FSDP training job, t5-3b.  The capture window included two batches … the associated screen shot focuses on the time to process one batch.  You can use NVIDIA’s nsys-ui to explore the data in the nsys profile report (unzip the file first).  The PCIE traffic is correlated in time with the python thread that is launching NCCL calls during the backward phase.  There is some PCIE activity for brief intervals during the forward phase, but that is insignificant compared to the traffic during the backward phase.  The FSDP job specified NCCL_ALGO=Ring and used just one network interface.  The ring topology used GPUs 0 and 1 for external network communication.  You can see that in the device monitor output dmon.txt, which is also attached.  The device monitor sampled all 8 GPUs on the master node at 1 sec intervals.  GPU 3 participates in the NCCL rings via NVLink.  We see significant PCIE traffic from host to device (pcirx) for all of the GPUs.  In this example the ratio of H2D traffic compared to D2H traffic is ~14x for the GPUs that are not sending or receiving data over the external network, namely GPUs 2,3,4,5,6,7.  The significant PCIE H2D traffic is not present if we simply run NCCL benchmarks (back-to-back NCCL calls of the same collectives and similar message sizes).

**Comment by raghukiran1224** [comment]
@awgu Pored over a bunch more runs and my hunch is as we increase compute communication overlap, this "problem" gets worse. Is there a way one can control amount of compute+comm overlap (some internal flag that we can tweak?)? This can help us debug the issue further.

**Comment by awgu** [comment]
For FSDP, I think if you set `backward_prefetch=None`, then there should not be communication/computation overlap in the backward pass.

**Comment by lchu-ibm** [comment]
@awgu my understanding is both `backward_prefetch` options (pre and post) would involve overlapping.
At any given time, there are three things can be done in parallel:
1. reduce-scatter comms for the **previous** block
2. gradient computation for the **current** block
3. all-gather comms for the **next** block

pre-fetch would overlap 2 and 3, while post-fetch would overlap 1 and 2.

**Comment by awgu** [comment]
@lchu-ibm I think for `BACKWARD_PRE`, we overlap current gradient computation with next all-gather, and for `BACKWARD_POST`, we overlap the current reduce-scatter with the next gradient computation.

However, if we set `backward_prefetch=None` (meaning neither `BACKWARD_PRE` nor `BACKWARD_POST`), then FSDP runs current gradient computation -> current reduce-scatter -> next all-gather -> next gradient computation -> next reduce-scatter etc, where each thing blocks the next, so there should be no overlap.

**Comment by lchu-ibm** [comment]
@awgu ah! I didn't realize we had the third mode - Neither.   I somehow thought we only had two modes which under the enum we set.  I will try the None mode. thanks for the info!

**Comment by lessw2020** [comment]
quick update on this:
a - worked with @lchu-ibm  today and verified this issue reproes on AWS, so it's pytorch specific and not their hardware issue.
b - tested this on both PT2 latest nightly and 1.13.1 - both show the same high traffic.
c - doubly confirmed that running the same toy model without DDP results in minimal traffic. 
<img width="612" alt="no_distributed" src="https://github.com/pytorch/pytorch/assets/46302957/85020073-2704-4c0d-a474-969cec1953f8">

**Comment by lchu-ibm** [comment]
providing a little more observations here:
1. high volume of traffic **is gone** when disabling overlapping on **nccl level**. (already shared above).
2. high volume of traffic **remains** when disabling overlapping on **torch level**.

- for 1, we did:  `CUDA_LAUNCH_BLOCKING=1`

- for 2, we did:

    - 2a: for DDP, we did `dist._DEFAULT_FIRST_BUCKET_BYTES = int(larger_number_here * 1024 * 1024)`

    - 2b: for FSDP, we did `backward_prefetch=None`

traces are attached as well. it seems all items are the same except the "time alignment" between cpu thread and gpu thread is different.  

[traces.zip](https://github.com/pytorch/pytorch/files/12101312/traces.zip)

**Comment by kwen2501** [comment]
I don't seem to be able to repro the issue on a different cluster, with a A100 + NVLink node. The rx and tx PCI-e traffic observed are 22 MB/s and 7 MB/s, respectively, per GPU. Likely due to input and output exchange between CPU and GPU. My PyTorch version is 2.1.0dev05262023 + CUDA 11.8.

**Comment by lchu-ibm** [comment]
sharing some updates here from various of offline discussions:

1. At this stage, we think this issue is likely a combination of our infra setup and PT, given the facts that:
    1.  This cannot be reproduced on a different cluster (Thanks to the help from @kwen2501 and @lessw2020). And upon initial check, there is no obvious difference in the nccl topo and both clusters use A100 + NVLink.  We will study the diff a little bit further by checking more carefully on the infra setup.
    2. This cannot be reproduced with PT version prior 2.0 with our setup (no traffic on DDP with v1.x, and much less traffic on FSDP with v1.x), exact same experiments and setup.

2. Latest observations (pretty much a summary of discussions in this thread):
    1. comms alone won't trigger the traffic.  we simulate the whole process but take away the computation part, and traffic is gone.
    2. compute alone won't trigger the traffic.  well, of course...
    3. extra traffic was only introduced when both computation and comms happened and overlapped
    4. disabling the overlapping between comms and compute on the nccl level would remove the extra traffic.
    5. disabling the overlapping between comms and compute on PT level would NOT remove the extra traffic.

**Comment by garrett361** [comment]
Hi all, great thread. I was wondering if there was an update to this issues. Does it persist today? Was it ever root-caused?  CC @lchu-ibm

**Comment by mori360** [comment]
We are committed to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
If you encounter any similar issues with FSDP2, please open a new issue with updated information.
